tests/utils/test_data_type_utils.py

```diff
@@ -0,0 +1,64 @@
+# SPDX-License-Identifier: Apache-2.0
+# https://spdx.dev/learn/handling-license-info/
+
+# Third Party
+import pytest
+import torch
+
+# Local
+from tuning.utils import data_type_utils
+
+dtype_dict = {
+    "bfloat16": torch.bfloat16,
+    "bits16": torch.bits16,
+    "bits1x8": torch.bits1x8,
+    "bits2x4": torch.bits2x4,
+    "bits4x2": torch.bits4x2,
+    "bits8": torch.bits8,
+    "bool": torch.bool,
+    "cdouble": torch.cdouble,
+    "cfloat": torch.cfloat,
+    "chalf": torch.chalf,
+    "complex128": torch.complex128,
+    "complex32": torch.complex32,
+    "complex64": torch.complex64,
+    "double": torch.double,
+    "float": torch.float,
+    "float16": torch.float16,
+    "float32": torch.float32,
+    "float64": torch.float64,
+    "float8_e4m3fn": torch.float8_e4m3fn,
+    "float8_e4m3fnuz": torch.float8_e4m3fnuz,
+    "float8_e5m2": torch.float8_e5m2,
+    "float8_e5m2fnuz": torch.float8_e5m2fnuz,
+    "half": torch.half,
+    "int": torch.int,
+    "int16": torch.int16,
+    "int32": torch.int32,
+    "int64": torch.int64,
+    "int8": torch.int8,
+    "long": torch.long,
+    "qint32": torch.qint32,
+    "qint8": torch.qint8,
+    "quint2x4": torch.quint2x4,
+    "quint4x2": torch.quint4x2,
+    "quint8": torch.quint8,
+    "short": torch.short,
+    "uint8": torch.uint8,
+}
+
+
+def test_str_to_torch_dtype():
+    for t in dtype_dict.keys():
+        assert data_type_utils.str_to_torch_dtype(t) == dtype_dict.get(t)
```

Comment:
im not sure if this is a good function to test. THis is a function that deals with configs, but for such things, its better to defer to huggingface, because changes come in very fast

As an example, this function will fail if `auto` is passed in, because it is not an "actual" torch dtype

however, it is a valid flag, as seen in this code https://github.com/huggingface/transformers/blob/main/src/transformers/modeling_utils.py#L3317-L3340

Reply:
>>> from tuning.utils import data_type_utils
>>>data_type_utils.get_torch_dtype("auto")
 ValueError: Unrecognized data type of a torch.Tensor

Reply:
right but what is the motivation of checking the formatting here, when it is already checked more throughly in the huggingface trainer? If there is an error, the huggingface trainer with throw the Exception which will be bubbled up the script. So what difference in behavior is this check trying to achieve, if in both cases, the idea is to return an error code.

Reply:
ok i see, I think i misunderstood https://github.com/huggingface/transformers/blob/0ad770c3733f9478a8d9d0bc18cc6143877b47a2/src/transformers/modeling_utils.py#L2707-L2727, 

the only `str` value it will take is `auto`, the rest relies on it to be a torch dtype. So i think its a valid function. 

If think omitting `auto` is a design decision. either way is fine.,

Reply:
Sorry an update to my previous comment
> the only str value it will take is auto, the rest relies on it to be a torch dtype. So i think its a valid function.

It turns out that `trl` has a `ModelConfig` dataclass which handles `torch_dtype`. I think its better to rely on that

https://github.com/huggingface/trl/blob/3bd02380c7e39cae2ec5013f27907b30bc4ce6bf/trl/trainer/model_config.py#L21

Reply:
@fabianlim Are you suggesting 
1. These functions need to to handle `auto`? If so create an issue.
2. Or we need to add more test case to test these functions? If so, what additional input to the functions do I need to test and what is the expected output of the functions do I need to assert.

Reply:
@tedhtchang as far as these unit go, they look great , thank you! we do not need to test each and every torch dtype , if conversion works for some , it will work for others. So i will add a separate comment on reducing the number of tests - remember every parameter value you pass is like another test - CI/CD tests also need to run fast. So we need CODE coverage and pass parameters that would go through a separate codepath while testing - example if "floats" were handled differently from "ints" then it makes sense to have 2 tests. Here all string to torch dtype is happening with 1 codepath - so we need not have 20+ tests for one codepath